src/agents/utils/signatures.ts

```diff
@@ -0,0 +1,139 @@
+import {
+  TypedDataDomain,
+  TypedDataField
+} from '@ethersproject/abstract-signer';
+import { Contract, ContractInterface } from '@ethersproject/contracts';
+import { _TypedDataEncoder } from '@ethersproject/hash';
+import { StaticJsonRpcProvider } from '@ethersproject/providers';
+import { verifyTypedData, Wallet } from '@ethersproject/wallet';
+
+type SignatureVerifier = (
+  address: string,
+  types: Record<string, TypedDataField[]>,
+  message: Record<string, any>,
+  signature: string
+) => Promise<boolean>;
+
+type VerifyOptions = {
+  /** Allow EIP-712 signatures. */
+  eip712?: boolean;
+  /** Allow EIP-1271 signatures. */
+  eip1271?: boolean;
+};
+
+const EVM_TYPED_DATA_CHAIN_ID = parseInt(
+  process.env.TYPED_DATA_CHAIN_ID ?? '11155111'
+);
+
+const ERC1271_ABI = [
+  'function isValidSignature(bytes32 _hash, bytes memory _signature) public view returns (bytes4 magicValue)'
+];
+const ERC1271_ABI_OLD = [
+  'function isValidSignature(bytes _hash, bytes memory _signature) public view returns (bytes4 magicValue)'
+];
+const ERC1271_MAGIC_VALUE = '0x1626ba7e';
+const ERC1271_MAGIC_VALUE_OLD = '0x20c13b0b';
+
+const DOMAIN: TypedDataDomain = {
+  name: 'highlight',
+  version: '0.1.0'
+};
+
+const provider = new StaticJsonRpcProvider(
+  `https://rpc.snapshot.org/${EVM_TYPED_DATA_CHAIN_ID}`,
+  EVM_TYPED_DATA_CHAIN_ID
+);
+
+function isEqual(a: string, b: string): boolean {
+  return a.toLowerCase() === b.toLowerCase();
+}
+
+function getHash(data: Record<string, any>): string {
+  const { types, message } = data;
+  return _TypedDataEncoder.hash(DOMAIN, types, message);
+}
+
+export async function signMessage(
+  wallet: Wallet,
+  types: Record<string, TypedDataField[]>,
+  message: Record<string, any>
+): Promise<string> {
+  return wallet._signTypedData(DOMAIN, types, message);
+}
+
+export const verifyEip712Signature: SignatureVerifier = async (
```

Comment:
How do you decide between `const` or `function`?

Reply:
Here arrow function is needed so we can use `SignatureVerifier` type annotation to get params and return value typed.